### PR TITLE
Setup terraform structure for connector dev infra for GCP and AWS

### LINKED ISSUE #3968: Version control integration test infra resources
## Tell us about the problem you're trying to solve
We are beginning to accumulate a non-trivial amount of infra used in integration tests: 
* RDS instance for MSSQL SSL (and soon for other DBs)
* BigQuery 
* Redshift
* Potentially SAP Hana
* etc...

We should use terraform to provision the resources used in testing. Or at least version control it somehow so it is clear how it's setup and how to access it/keep track of it. 

## Describe the solution you’d like
Terraform or some other form of version-controllable way of keeping track of these infra resources

### LINKED ISSUE #5166: Split monolithic connector dev terraform state
## Tell us about the problem you're trying to solve
Our current terraform project introduced in https://github.com/airbytehq/airbyte/pull/4641 uses a monolithic structure to define resources meaning there is a single entrypoint where `terraform plan/apply/etc.`  are run. This is not great for isolation of resources since a rogue `terraform destroy` could wreck resources unintentionally. 

We should split these resources such that logically distinct units cannot destroy each other.

### PR DESCRIPTION
**Pull request by sherifnada**
## What
adds basic TF structure for adding terraform infra. See the README to understand its usage. 

Goals of this setup: 
* Autonomy/self-serve for the developer: I don’t need to talk to anyone to spin up my resources
* isolation of resources: i can create whatever I want and not worry that I’m screwing up someone else’s test infra

## How
* All connector-infra terraform code lives in `terraform/<cloud_provider>/connectors`. A connector's infra goes in `terraform/<cloud_provider>/connectors/<connector-name>`. 
  * sidenote: I'm not exactly sure how much infra connectors should share. Rule of thumb i have in mind is each connector's infra should be completely isolated unless there is a good reason to do otherwise (price, complex setups which benefit from DRY, or project-level resource quotas). I do NOT have great guidelines on the tipping point for these at the moment. It is left to the developer's judgment. 
* The backend for each will be hosted on that cloud (GCP backend is GCS, AWS is S3, etc..). 
* A developer who needs to edit infra needs to:
  *  follows the instructions in the README to obtain credentials for the relevant backend
  * edit TF code and apply it 
  * create a PR containing the TF plan
  
I didn't choose to use a hosted solution like Terraform Cloud because we have a lot of devs who will be touching this infra which would be both expensive and high friction to manage users. 

## Recommended reading order
1. `terraform/README.md`
2. files in `terraform/aws/`

Relates to #4641 
Closes #5166

### CODE REVIEW

File: terraform/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+To run terraform commands use the TF wrapper: 
+
+```
+# From the airbyte repo root
+./tools/terraform/terraform.sh ./terraform/<cloud-provider>/connectors/ <TF command>
+```
+
+_hint_: alias `atf` (short for `airbyte-terraform`) to `./tools/terraform/terraform.sh` to 10x your iteration speed.    
+
+**If this is your first time running Terraform** run the `init` command
```
Comment (jrhizor) [inline terraform/README.md:33]: For all of these commands please give the full command so there's no ambiguity `./tools/terraform/terraform.sh init`, `./tools/terraform/terraform.sh apply`, etc. It also makes it copy-pastable.

**Review by jrhizor — APPROVED** [review]

File: tools/terraform/terraform.sh [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env bash
+# This file runs the Terraform CLI from a container to ensure repeatable results
+
+set -e
+
```
Comment (davinchia) [inline tools/terraform/terraform.sh:5]: this is the file that will be replaced by tfenv?

**Review by davinchia — APPROVED** [review]

**Review by airbyte-jenny — COMMENTED** [review]
PR has been updated with AWS terraform resources, for speed of development.  We'll likely move these to GCP when we're no longer under deadline and I can tackle the learning curve.  The infrastructure is all set up directly from terraform except for the postgres user, password, and grants.  Those must be done in a sql client by a human for now.  Plan contents attached below.

shared/* - This bootstraps the terraform backend for this aws subaccount.  This is a kms key for encryption, with an alias terraform_s3_kms.  It's a dynamodb table terraform-state-lock-dynamo used for locking terraform runs.  It's an S3 bucket com-airbyte-terraform-state with versioning and encryption at rest, to store the remote terraform state files safely even when they contain secrets.  The bootstrap process is tedious as you must use local terraform without a lock file to create the resources, and then migrate the backend to the remote state backend.  Redoing this to get a 'terraform plan' output would not be especially valuable.

ssh_tunnel/* - This area creates all of the integration test infrastructure for the ssh_tunnel project.  It relies on the existing terraform backend definition to supply locking and remote state.  Only the SQL user creation and grants are done by hand.

**Comment by airbyte-jenny** [comment]
Terraform plan output.

`

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # aws_key_pair.key will be created
  + resource "aws_key_pair" "key" {
      + arn         = (known after apply)
      + fingerprint = (known after apply)
      + id          = (known after apply)
      + key_name    = "dbtunnel-bastion-ec2-user-ssh-key"
      + key_pair_id = (known after apply)
      + public_key  = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQCtNfcjnlbnmMXcYuscE96X3jp6uXbcjMFeloBPpeV5r9WwuAcczo7Sm8gqUBDtQrJ/TtGTZ+plAX5w8RlbQ7NsWoOIMhHaKmR/USo2kxiwXfQ3J8nJH42lGT/1lDKR01q3rQvjSNF4sJ4uJ7Bb7XWWLnT4X0IxCHrmNAEX/TH0D8RByFbjiHmw0avsHBt58Qux7AtAgZ4zqN0Q4NI3BX7uK3hti9H2I/QM3GYIDJSzcas6PUKxSh/QmaGfae0UBhavE4mrYXwftK6Q2lL7flU8b01eL9q+sfS5ncMIE7FA99+fXFB3vm3INNZysaMUGh/cCS+5vGEm2kD/T/ZZWAMIngjr6KRSUAvinSKMfw1aILm0dpQgty3E9lXXJm4cE49jS/tull6didwwWEvtchYHZIGSz+F7yuLswukz8hNC2y1wKS8046yYgpdO2C41HcUpfJyQ7YmR5Q40FOJjXS9twxCSvbsDr/DfRqA11rPuXt0VwBKPAFWzonJ0jCSnrt8= ec2-user"
      + tags_all    = (known after apply)
    }

  # module.ssh_tunnel_testing.aws_db_instance.default will be created
  + resource "aws_db_instance" "default" {
      + address                               = (known after apply)
      + allocated_storage                     = 5
      + apply_immediately                     = true
      + arn                                   = (known after apply)
      + auto_minor_version_upgrade            = true
      + availability_zone                     = (known after apply)
      + backup_retention_period               = (known after apply)
      + backup_window                         = (known after apply)
      + ca_cert_identifier                    = (known after apply)
      + character_set_name                    = (known after apply)
      + copy_tags_to_snapshot                 = false
      + db_subnet_group_name                  = "dbtunnel-public-dbsubnet-group"
      + delete_automated_backups              = true
      + endpoint                              = (known after apply)
      + engine                                = "postgres"
      + engine_version                        = "12.6"
      + hosted_zone_id                        = (known after apply)
      + id                                    = (known after apply)
      + identifier                            = "tunnel-dev"
      + identifier_prefix                     = (known after apply)
      + instance_class                        = "db.t3.small"
      + kms_key_id                            = (known after apply)
      + latest_restorable_time                = (known after apply)
      + license_model                         = (known after apply)
      + maintenance_window                    = (known after apply)
      + monitoring_interval                   = 0
      + monitoring_role_arn                   = (known after apply)
      + multi_az                              = (known after apply)
      + name                                  = "airbyte"
      + option_group_name                     = (known after apply)
      + parameter_group_name                  = "rds-pg"
      + password                              = (sensitive value)
      + performance_insights_enabled          = false
      + performance_insights_kms_key_id       = (known after apply)
      + performance_insights_retention_period = (known after apply)
      + port                                  = (known after apply)
      + publicly_accessible                   = false
      + replicas                              = (known after apply)
      + resource_id                           = (known after apply)
      + skip_final_snapshot                   = true
      + snapshot_identifier                   = (known after apply)
      + status                                = (known after apply)
      + storage_type                          = (known after apply)
      + tags_all                              = (known after apply)
      + timezone                              = (known after apply)
      + username                              = "airbyte"
      + vpc_security_group_ids                = (known after apply)
    }

  # module.ssh_tunnel_testing.aws_db_parameter_group.default will be created
  + resource "aws_db_parameter_group" "default" {
      + arn         = (known after apply)
      + description = "RDS default parameter group"
      + family      = "postgres12"
      + id          = (known after apply)
      + name        = "rds-pg"
      + name_prefix = (known after apply)
      + tags_all    = (known after apply)
    }

  # module.ssh_tunnel_testing.aws_db_subnet_group.default will be created
  + resource "aws_db_subnet_group" "default" {
      + arn         = (known after apply)
      + description = "Managed by Terraform"
      + id          = (known after apply)
      + name        = "dbtunnel-public-dbsubnet-group"
      + name_prefix = (known after apply)
      + subnet_ids  = (known after apply)
      + tags        = {
          + "Name" = "dbtunnel-public-dbsubnet-group"
        }
      + tags_all    = {
          + "Name" = "dbtunnel-public-dbsubnet-group"
        }
    }

  # module.ssh_tunnel_testing.aws_eip.dbtunnel-eip will be created
  + resource "aws_eip" "dbtunnel-eip" {
      + allocation_id        = (known after apply)
      + association_id       = (known after apply)
      + carrier_ip           = (known after apply)
      + customer_owned_ip    = (known after apply)
      + domain               = (known after apply)
      + id                   = (known after apply)
      + instance             = (known after apply)
      + network_border_group = (known after apply)
      + network_interface    = (known after apply)
      + private_dns          = (known after apply)
      + private_ip           = (known after apply)
      + public_dns           = (known after apply)
      + public_ip            = (known after apply)
      + public_ipv4_pool     = (known after apply)
      + tags_all             = (known after apply)
      + vpc                  = true
    }

  # module.ssh_tunnel_testing.aws_eip_association.dbtunnel-eip-assoc will be created
  + resource "aws_eip_association" "dbtunnel-eip-assoc" {
      + allocation_id        = (known after apply)
      + id                   = (known after apply)
      + instance_id          = (known after apply)
      + network_interface_id = (known after apply)
      + private_ip_address   = (known after apply)
      + public_ip            = (known after apply)
    }

  # module.ssh_tunnel_testing.aws_instance.dbtunnel-bastion will be created
  + resource "aws_instance" "dbtunnel-bastion" {
      + ami                                  = "ami-0233c2d874b811deb"
      + arn                                  = (known after apply)
      + associate_public_ip_address          = (known after apply)
      + availability_zone                    = (known after apply)
      + cpu_core_count                       = (known after apply)
      + cpu_threads_per_core                 = (known after apply)
      + get_password_data                    = false
      + host_id                              = (known after apply)
      + id                                   = (known after apply)
      + instance_initiated_shutdown_behavior = (known after apply)
      + instance_state                       = (known after apply)
      + instance_type                        = "t3.small"
      + ipv6_address_count                   = (known after apply)
      + ipv6_addresses                       = (known after apply)
      + key_name                             = "dbtunnel-bastion-ec2-user-ssh-key"
      + outpost_arn                          = (known after apply)
      + password_data                        = (known after apply)
      + placement_group                      = (known after apply)
      + primary_network_interface_id         = (known after apply)
      + private_dns                          = (known after apply)
      + private_ip                           = (known after apply)
      + public_dns                           = (known after apply)
      + public_ip                            = (known after apply)
      + secondary_private_ips                = (known after apply)
      + security_groups                      = (known after apply)
      + source_dest_check                    = true
      + subnet_id                            = (known after apply)
      + tags                                 = {
          + "Name" = "dbtunnel-bastion"
        }
      + tags_all                             = {
          + "Name" = "dbtunnel-bastion"
        }
      + tenancy                              = (known after apply)
      + user_data                            = "9a9b3bb0f1cfa9adbe7636cff1da40a10353e82b"
      + vpc_security_group_ids               = (known after apply)

      + capacity_reservation_specification {
          + capacity_reservation_preference = (known after apply)

          + capacity_reservation_target {
              + capacity_reservation_id = (known after apply)
            }
        }

      + ebs_block_device {
          + delete_on_termination = (known after apply)
          + device_name           = (known after apply)
          + encrypted             = (known after apply)
          + iops                  = (known after apply)
          + kms_key_id            = (known after apply)
          + snapshot_id           = (known after apply)
          + tags                  = (known after apply)
          + throughput            = (known after apply)
          + volume_id             = (known after apply)
          + volume_size           = (known after apply)
          + volume_type           = (known after apply)
        }

      + enclave_options {
          + enabled = (known after apply)
        }

      + ephemeral_block_device {
          + device_name  = (known after apply)
          + no_device    = (known after apply)
          + virtual_name = (known after apply)
        }

      + metadata_options {
          + http_endpoint               = (known after apply)
          + http_put_response_hop_limit = (known after apply)
          + http_tokens                 = (known after apply)
        }

      + network_interface {
          + delete_on_termination = (known after apply)
          + device_index          = (known after apply)
          + network_interface_id  = (known after apply)
        }

      + root_block_device {
          + delete_on_termination = (known after apply)
          + device_name           = (known after apply)
          + encrypted             = (known after apply)
          + iops                  = (known after apply)
          + kms_key_id            = (known after apply)
          + tags                  = (known after apply)
          + throughput            = (known after apply)
          + volume_id             = (known after apply)
          + volume_size           = (known after apply)
          + volume_type           = (known after apply)
        }
    }

  # module.ssh_tunnel_testing.aws_route_table.dbtunnel-public-route will be created
  + resource "aws_route_table" "dbtunnel-public-route" {
      + arn              = (known after apply)
      + id               = (known after apply)
      + owner_id         = (known after apply)
      + propagating_vgws = (known after apply)
      + route            = [
          + {
              + carrier_gateway_id         = ""
              + cidr_block                 = "0.0.0.0/0"
              + destination_prefix_list_id = ""
              + egress_only_gateway_id     = ""
              + gateway_id                 = "igw-03f43f8e7909d7975"
              + instance_id                = ""
              + ipv6_cidr_block            = ""
              + local_gateway_id           = ""
              + nat_gateway_id             = ""
              + network_interface_id       = ""
              + transit_gateway_id         = ""
              + vpc_endpoint_id            = ""
              + vpc_peering_connection_id  = ""
            },
        ]
      + tags             = {
          + "Name" = "dbtunnel-public-route"
        }
      + tags_all         = {
          + "Name" = "dbtunnel-public-route"
        }
      + vpc_id           = "vpc-001ad881b80193126"
    }

  # module.ssh_tunnel_testing.aws_route_table_association.dbtunnel-route-assoc-private-subnet-2 will be created
  + resource "aws_route_table_association" "dbtunnel-route-assoc-private-subnet-2" {
      + id             = (known after apply)
      + route_table_id = (known after apply)
      + subnet_id      = (known after apply)
    }

  # module.ssh_tunnel_testing.aws_route_table_association.dbtunnel-route-assoc-public-subnet-1 will be created
  + resource "aws_route_table_association" "dbtunnel-route-assoc-public-subnet-1" {
      + id             = (known after apply)
      + route_table_id = (known after apply)
      + subnet_id      = (known after apply)
    }

  # module.ssh_tunnel_testing.aws_security_group.dbtunnel-sg will be created
  + resource "aws_security_group" "dbtunnel-sg" {
      + arn                    = (known after apply)
      + description            = "Allow inbound traffic but only from the dbtunnel subnet"
      + egress                 = [
          + {
              + cidr_blocks      = [
                  + "0.0.0.0/0",
                ]
              + description      = ""
              + from_port        = 0
              + ipv6_cidr_blocks = [
                  + "::/0",
                ]
              + prefix_list_ids  = []
              + protocol         = "-1"
              + security_groups  = []
              + self             = false
              + to_port          = 0
            },
        ]
      + id                     = (known after apply)
      + ingress                = [
          + {
              + cidr_blocks      = [
                  + "10.0.40.0/24",
                ]
              + description      = "tcp on 5432 from subnet"
              + from_port        = 5432
              + ipv6_cidr_blocks = []
              + prefix_list_ids  = []
              + protocol         = "tcp"
              + security_groups  = []
              + self             = false
              + to_port          = 5432
            },
        ]
      + name                   = "dbtunnel-sg-allow-postgres"
      + name_prefix            = (known after apply)
      + owner_id               = (known after apply)
      + revoke_rules_on_delete = false
      + tags                   = {
          + "Name" = "dbtunnel-sg-allow-postgres"
        }
      + tags_all               = {
          + "Name" = "dbtunnel-sg-allow-postgres"
        }
      + vpc_id                 = "vpc-001ad881b80193126"
    }

  # module.ssh_tunnel_testing.aws_security_group.ssh-and-egress-allowed will be created
  + resource "aws_security_group" "ssh-and-egress-allowed" {
      + arn                    = (known after apply)
      + description            = "Managed by Terraform"
      + egress                 = [
          + {
              + cidr_blocks      = [
                  + "0.0.0.0/0",
                ]
              + description      = ""
              + from_port        = 0
              + ipv6_cidr_blocks = []
              + prefix_list_ids  = []
              + protocol         = "-1"
              + security_groups  = []
              + self             = false
              + to_port          = 0
            },
        ]
      + id                     = (known after apply)
      + ingress                = [
          + {
              + cidr_blocks      = [
                  + "0.0.0.0/0",
                ]
              + description      = ""
              + from_port        = 22
              + ipv6_cidr_blocks = []
              + prefix_list_ids  = []
              + protocol         = "tcp"
              + security_groups  = []
              + self             = false
              + to_port          = 22
            },
        ]
      + name                   = (known after apply)
      + name_prefix            = (known after apply)
      + owner_id               = (known after apply)
      + revoke_rules_on_delete = false
      + tags                   = {
          + "Name" = "ssh-and-egress-allowed"
        }
      + tags_all               = {
          + "Name" = "ssh-and-egress-allowed"
        }
      + vpc_id                 = "vpc-001ad881b80193126"
    }

  # module.ssh_tunnel_testing.aws_subnet.main-subnet-private-dbtunnel will be created
  + resource "aws_subnet" "main-subnet-private-dbtunnel" {
      + arn                             = (known after apply)
      + assign_ipv6_address_on_creation = false
      + availability_zone               = "us-east-2b"
      + availability_zone_id            = (known after apply)
      + cidr_block                      = "10.0.41.0/24"
      + id                              = (known after apply)
      + ipv6_cidr_block_association_id  = (known after apply)
      + map_public_ip_on_launch         = false
      + owner_id                        = (known after apply)
      + tags                            = {
          + "Name" = "private-dbtunnel"
        }
      + tags_all                        = {
          + "Name" = "private-dbtunnel"
        }
      + vpc_id                          = "vpc-001ad881b80193126"
    }

  # module.ssh_tunnel_testing.aws_subnet.main-subnet-public-dbtunnel will be created
  + resource "aws_subnet" "main-subnet-public-dbtunnel" {
      + arn                             = (known after apply)
      + assign_ipv6_address_on_creation = false
      + availability_zone               = "us-east-2a"
      + availability_zone_id            = (known after apply)
      + cidr_block                      = "10.0.40.0/24"
      + id                              = (known after apply)
      + ipv6_cidr_block_association_id  = (known after apply)
      + map_public_ip_on_launch         = true
      + owner_id                        = (known after apply)
      + tags                            = {
          + "Name" = "public-dbtunnel"
        }
      + tags_all                        = {
          + "Name" = "public-dbtunnel"
        }
      + vpc_id                          = "vpc-001ad881b80193126"
    }

Plan: 14 to add, 0 to change, 0 to destroy.

─────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't
guarantee to take exactly these actions if you run "terraform apply" now.

`

File: docs/SUMMARY.md [hunk 3]
```diff
@@ -179,10 +200,10 @@
   * [License](project-overview/license.md)
 * [Careers & Open Positions](career-and-open-positions/README.md)
   * [Senior Software Engineer](career-and-open-positions/senior-software-engineer.md)
-  * [Founding Developer Evangelist](career-and-open-positions/founding-developer-evangelist.md)
-  * [Senior Technical Writer / Editor in Chief](career-and-open-positions/technical-content-writer.md)
-  * [Senior Technical Recruiter](career-and-open-positions/recruiter.md)
-  * [Head of Lead Generation](career-and-open-positions/head-of-lead-generation.md)
+  * [Senior Product Manager](career-and-open-positions/senior-product-manager.md)
```
Comment (sherifnada) [inline docs/SUMMARY.md:203]: is this up to date with master? just asking since these changes are unrelated 🤔

File: terraform/README.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+## Connector Development Infrastructure
+We use Terraform to manage any persistent infrastructure used for developing or testing connectors.
+
+Directory structure is roughly as follows:
+
+    ├── aws
+    │   ├── demo
+    │   │   ├── core
+    │   │   └── lb
+    │   ├── shared
```
Comment (sherifnada) [inline terraform/README.md:12]: Should we move all connector-related infra to `airbyte-integrations/infra`? Probably better to scope it that way but if it causes disruption then we don't need to

File: docs/SUMMARY.md (hunk 3 above)
Comment (airbyte-jenny) [reply docs/SUMMARY.md:203]: I grabbed it from master because of merge conflicts that seemed to be irrelevant.  Looked like just a tiny local edit happened in our branch, causing conflict.  I kept what was in master.

File: terraform/README.md (hunk 4 above)
Comment (airbyte-jenny) [reply terraform/README.md:12]: Yeah, I agree.  Moved.
